### PR TITLE
Immutable b tree

### PR DESCRIPTION
**Pull request by dustinhiatt-wf**
Adds an immutable btree to the btree package.

@Workiva/dtbackend

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -145,6 +145,9 @@ immutable version.  Unfortunately, to make the B-tree generic we require an
 interface and the most expensive operation in CPU profiling is the interface
 method which in turn calls into runtime.assertI2T.  We need generics.
 
+#### Immutable B Tree
+A btree based on two principals, immutablability and concurrency.  Somewhat slow for single value lookups and puts, it is very fast for bulk operations.  A persister can be injected to make this index persistent.
```
Comment (alexandercampbell-wf) [inline README.md]: This should probably wrap at 80 chars to fit with the other entries in the README.

File: btree/immutable/cacher.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
+package btree
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/futures"
+)
+
+// cacher provides a convenient interface for retrieving,
```
Comment (alexandercampbell-wf) [inline btree/immutable/cacher.go]: `cacher` is not quite an interface... Maybe you could say in this docstring that it's added threadsafety and caching around a Persister?

File: btree/immutable/cacher.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+	}
+
+	return n, nil
+}
+
+// getNode will return a Node matching the provided id.  An error is returned
+// if the cacher could not go to the persister or the node could not be found.
+// All found nodes are cached so subsequent calls should be faster than
+// the initial.  This blocks until the node is loaded, but is also threadsafe.
+func (c *cacher) getNode(t *Tr, key ID, cache bool) (*Node, error) {
```
Comment (alexandercampbell-wf) [inline btree/immutable/cacher.go]: Personally, I'd rename the `cache` variable to `useCache` to more clearly express that it's a boolean and not an object.

File: btree/immutable/cacher.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+			return
+		}
+
+		if n == nil {
+			completer <- ErrNodeNotFound
+			return
+		}
+
+		completer <- n
+	}()
```
Comment (alexandercampbell-wf) [inline btree/immutable/cacher.go]: Could this be made into a method on `cacher`? That way you could just write

```go
go c.asyncLoadNode(t, key, completer)
```

File: btree/immutable/config.go [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	// NodeWidth defines the branching factor of the tree.  Any node
+	// wider than this value will get split and when the width of a node
+	// falls to less than half this value the node gets merged.  This
+	// ensures optimal performance while running to the key value store.
+	NodeWidth int
+	// Perister defines the key value store that the tree can use to
+	// save and load nodes.
+	Persister Persister
+	// Comparator is the function used to determine ordering.
+	Comparator Comparator `msg:"-"`
```
Comment (alexandercampbell-wf) [inline btree/immutable/config.go]: Why would this be `msg:"-"`?

File: btree/immutable/node_gen.go [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+	for cmr := range z.ChildValues {
+		o, err = msgp.AppendIntf(o, z.ChildValues[cmr])
+		if err != nil {
+			return
+		}
+	}
+	// string "ck"
+	o = append(o, 0xa2, 0x63, 0x6b)
+	o = msgp.AppendArrayHeader(o, uint32(len(z.ChildKeys)))
+	for ajw := range z.ChildKeys {
```
Comment (alexandercampbell-wf) [inline btree/immutable/node_gen.go]: Would you mind upgrading to the latest `msgp`? This generated code looks like it was generated by an older version.

File: btree/immutable/config.go (hunk 5 above)
Comment (dustinhiatt-wf) [inline btree/immutable/config.go]: It's a function.

File: btree/immutable/node_gen.go (hunk 6 above)
Comment (dustinhiatt-wf) [inline btree/immutable/node_gen.go]: Done.

Comment (alexandercampbell-wf) [inline btree/immutable/node_gen.go]: Not seeing the commit; can you push the regeneration?

**Comment by alexandercampbell-wf** [comment]
+1

**Pull request merged**